packages/loader/container-loader/src/test/serializedStateManager.spec.ts

```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
 		);
-		// equivalent to attach
-		serializedStateManager.setSnapshot({
-			baseSnapshot: { trees: {}, blobs: {} },
-			snapshotBlobs: {},
-		});
-		for (let num = 0; num < 10; ++num) {
-			serializedStateManager.addProcessedOp(generateSavedOp());
-		}
-		await serializedStateManager.getPendingLocalStateCore(
+		const { baseSnapshot, version } = await serializedStateManager.fetchSnapshot(
+			undefined,
+			false,
+		);
+		assert(baseSnapshot);
+		assert.strictEqual(version?.id, "fromStorage");
+		assert.strictEqual(version.treeId, "fromStorage");
+		const state = await serializedStateManager.getPendingLocalStateCore(
 			{ notifyImminentClosure: false },
 			"clientId",
 			new MockRuntime(),
 			resolvedUrl,
 		);
+		const parsed = JSON.parse(state);
+		assert.strictEqual(parsed.baseSnapshot.id, "fromStorage");
+		const attributes = getAttributesFromPendingState(parsed);
+		assert.strictEqual(attributes.sequenceNumber, 0);
+		assert.strictEqual(attributes.minimumSequenceNumber, 0);
 	});
 
-	it("can fetch snapshot and get state from it", async () => {
+	it("fetched snapshot is the same as pending snapshot", async () => {
+		const pending: IPendingContainerState = {
+			...pendingLocalState,
+			baseSnapshot: { ...snapshot, id: "fromPending" },
+		};
 		const storageAdapter = new MockStorageAdapter();
+		const deferred = new Deferred<void>();
+		const cb = () => {
+			deferred.resolve();
+		};
 		const serializedStateManager = new SerializedStateManager(
-			undefined,
-			logger,
+			pending,
+			logger.toTelemetryLogger(),
 			storageAdapter,
 			true,
+			cb,
 		);
+		for (let num = 0; num < 10; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
 		const { baseSnapshot, version } = await serializedStateManager.fetchSnapshot(
 			undefined,
 			false,
 		);
-		assert(baseSnapshot);
-		assert.strictEqual(version?.id, "test");
-		assert.strictEqual(version.treeId, "test");
+		assert.strictEqual(baseSnapshot.id, "fromPending");
+		assert.strictEqual(version, undefined);
+		await deferred.promise;
+		logger.assertMatchAny([
+			{
+				category: "generic",
+				eventName: "serializedStateManager:OldSnapshotFetchWhileRefreshing",
+				snapshotSequenceNumber: 0,
+				firstProcessedOpSequenceNumber: 1,
+			},
+		]);
 		const state = await serializedStateManager.getPendingLocalStateCore(
 			{ notifyImminentClosure: false },
 			"clientId",
 			new MockRuntime(),
 			resolvedUrl,
 		);
 		const parsed = JSON.parse(state);
-		assert.strictEqual(parsed.baseSnapshot.id, "SnapshotId");
+		// We keep using the pending snapshot since it is the same based on its sequence number 0
+		assert.strictEqual(parsed.baseSnapshot.id, "fromPending");
 		const attributes = getAttributesFromPendingState(parsed);
 		assert.strictEqual(attributes.sequenceNumber, 0);
 		assert.strictEqual(attributes.minimumSequenceNumber, 0);
 	});
+
+	it("snapshot is older than first processed op", async () => {
+		const pending: IPendingContainerState = {
```

Comment:
i would like to see some tests where the pending state contains saved ops before the refresh